### PR TITLE
Fix Sorting bug

### LINKED ISSUE #61: Bug: Sort feature does not handle non-alphanumeric tag argument
### Bug found during testing
The input `sort t/!!friend` does not give an appropriate warning to users.
The reason seems to be that the tags cannot take non-alphanumeric arguments.

### Attached Screenshot of Bug

image

### PR DESCRIPTION
**Pull request by Eugene-Ong-W-X**
Non-alphanumeric tag names are now properly caught. 
Tag error message is displayed in response.
3 Junit tests re-enabled.

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java [hunk 1]
```diff
@@ -64,11 +72,15 @@ private SortArgument convertArguments(PrefixArgument prefixArg) {
...
-            return new SortArgument(prefix, isReverse, null);
+        try {
+            if (prefix.equals(PREFIX_TAG)) {
+                Tag tag = isReverse ? parseTag(arg.substring(1)) : parseTag(arg);
+                return new SortArgument(prefix, isReverse, tag);
+            } else {
+                return new SortArgument(prefix, isReverse, null);
+            }
+        } catch (ParseException e) {
+            return null;
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/logic/parser/SortCommandParser.java:83]: If I recall correctly, returning `null` is not a good practice.

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java [hunk 2]
```diff
@@ -37,6 +38,13 @@ public SortCommand parse(String args) throws ParseException {
         }
 
         List<SortArgument> sortArgs = convertPrefixArgToSortArg(argList);
+
+        for (SortArgument arg : sortArgs) {
+            if (arg == null) {
+                throw new ParseException(Tag.MESSAGE_CONSTRAINTS);
+            }
```
Comment (PangKuangWei) [inline src/main/java/seedu/address/logic/parser/SortCommandParser.java:45]: Should empty strings be guarded against as well?

**Review by PangKuangWei — CHANGES_REQUESTED** [review]
There are some considerations I would like you to consider for this bug fix.

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java (hunk 1 above)
Comment (Eugene-Ong-W-X) [reply src/main/java/seedu/address/logic/parser/SortCommandParser.java:83]: Both `convertArguments`, and `convertPrefixArgToSortArg` that calls it are private methods. 
Hence, there is no risk of the null objects returned triggering a `NullPointerException` outside of this class. 

As for the `parse` method, right after `convertPrefixArgToSortArg` is called, the returned list is checked for `null` objects.
Thus, the `null` here should be safe to use.

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java (hunk 2 above)
Comment (Eugene-Ong-W-X) [reply src/main/java/seedu/address/logic/parser/SortCommandParser.java:45]: Empty strings are guarded against in the `parseTag` method which is called in `convertArguments`. Hence, the `null` object here actually guards against **all** invalid tag names.

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java (hunk 1 above)
Comment (PangKuangWei) [reply src/main/java/seedu/address/logic/parser/SortCommandParser.java:83]: Ok

File: src/main/java/seedu/address/logic/parser/SortCommandParser.java (hunk 2 above)
Comment (PangKuangWei) [reply src/main/java/seedu/address/logic/parser/SortCommandParser.java:45]: Oh I see. Okay.

**Review by PangKuangWei — APPROVED** [review]

**Pull request merged**